src/StackExchange.Redis/PhysicalConnection.cs

```diff
@@ -1696,11 +1729,55 @@ private void MatchResult(in RawResult result)
             if (msg.ComputeResult(this, result))
             {
                 _readStatus = msg.ResultBoxIsAsync ? ReadStatus.CompletePendingMessageAsync : ReadStatus.CompletePendingMessageSync;
-                msg.Complete();
+                if (!msg.IsHighIntegrity)
+                {
+                    // can't complete yet if needs checksum
+                    msg.Complete();
+                }
             }
+            if (msg.IsHighIntegrity)
+            {
+                // stash this for the next non-OOB response
+                Volatile.Write(ref _awaitingChecksum, msg);
+            }
+
+
             _readStatus = ReadStatus.MatchResultComplete;
             _activeMessage = null;
 
+            static void ProcessHighIntegrityResponseChecksum(Message message, in RawResult result, PhysicalBridge? bridge)
+            {
+                bool isValid = false;
+                if (result.Resp2TypeBulkString == ResultType.BulkString)
+                {
+                    var payload = result.Payload;
+                    if (payload.Length == 4)
+                    {
+                        int interpreted;
+                        if (payload.IsSingleSegment)
+                        {
+                            interpreted = BinaryPrimitives.ReadInt32LittleEndian(payload.First.Span);
```

Comment:
Diccussed: might be we can simplify if we store the expected sequence on the `Message` do we can check/mate it when dequeue from _writtenAwaitingResponse and then we don't need to track state anywhere else

Reply:
we literally do exactly that already - `Message._highIntegrityToken` (note renamed slightly) - however, the problem is that *currently* the read loop only handles a single inbound at a time, and would require a significant refactor; the "move all the things" branch has scope to do better here